### PR TITLE
Sync gift_aid date changes.

### PR DESCRIPTION
**Pull request by ivan-unfolds**

### CODE REVIEW

File: api/models/Members.js [hunk 1]
```diff
@@ -47,6 +47,22 @@ var handle_membership_change = (member, cb) =>
...
+    .exec(function (error, item) {
+      if (error) return sails.log.error(error)
+      if (member.gift_aid_signed === item.gift_aid_signed) return cb(member)
+      if (member.gift_aid_signed) {
+        member.date_gift_aid_signed = new Date()
+        member.date_gift_aid_cancelled = null
+        return cb(member)
+      }
+      member.date_gift_aid_cancelled = new Date()
+      member.date_gift_aid_signed = null
```
Comment (rjmk) [inline api/models/Members.js:62]: Did we want to do this?

**Comment by ivan-unfolds** [comment]
@rjmk Working on syncing dates for Gift Aid. I did this on the front end on the member's view but have realised that we have to do it on admin view as well so think a solution on the back end is best.

I am also still a little confused on `gift_aid_signed` and `gift_aid_cancelled` fields. From what I can see we do not display `gift_aid_cancelled` anywhere, neither is there any option for an admin or a member to toggle this field themselves. I have an urge to just drop `gift_aid_cancelled` (as in not use it). I am not sure if Richard cares about the history of this field. I think he just cares about the current state of it, that is if `gift_aid_signed` is `false`, when was it revoked, and if it is `true` when was it set signed. What do you think?

File: api/models/Members.js (hunk 1 above)
Comment (ivan-unfolds) [reply api/models/Members.js:62]: Sorry should have assigned you after my comment below.

File: api/models/Members.js [hunk 2]
```diff
@@ -257,5 +273,5 @@ module.exports = {
   // ------------------------------------------------------------
   },
   beforeCreate: hash_password('password'),
-  beforeUpdate: (member, cb) => handle_membership_change(member, member => hash_password('new_password')(member, cb))
+  beforeUpdate: (member, cb) => handle_membership_change(member, member => handle_gift_aid_change(member, member => hash_password('new_password')(member, cb)))
```
Comment (rjmk) [inline api/models/Members.js]: Let's change the signatures to `cb => member` and then we can do

``` js
beforeCreate: (member, cb) => hash_password('password')(cb)(member),
beforeUpdate: (member, cb) => handle_membership_change(handle_gift_aid_change(hash_password('new_password')(cb)(member)))
```

**Comment by ivan-unfolds** [comment]
@rjmk made a few changes. What do you think?

**Comment by rjmk** [comment]
I like the changes. I think you might be right about not needing cancelled. Would you be OK to check in with Richard about that?

**Comment by ivan-unfolds** [comment]
Yup I will email him.

**Comment by rjmk** [comment]
Please do it by Pivotal if possible!

**Comment by ivan-unfolds** [comment]
👍

**Comment by ivan-unfolds** [comment]
@rjmk Richard has confirmed that we can drop `gift_aid_cancelled`.

File: api/models/Members.js [hunk 3]
```diff
@@ -256,6 +272,7 @@ module.exports = {
     }
   // ------------------------------------------------------------
   },
-  beforeCreate: hash_password('password'),
-  beforeUpdate: (member, cb) => handle_membership_change(member, member => hash_password('new_password')(member, cb))
+  beforeCreate: (member, cb) => hash_password('password')(cb)(member),
+  // beforeUpdate: (member, cb) => handle_membership_change(member, member => handle_gift_aid_change(member, member => hash_password('new_password')(member, cb)))
```
Comment (rjmk) [inline api/models/Members.js]: We should drop this comment

**Comment by rjmk** [comment]
Can you update the model and the blueprint to do away with `gift_aid_cancelled`?

**Comment by ivan-unfolds** [comment]
Leave it in my capable hands Rafe.

**Comment by rjmk** [comment]
Very nice! Capable 👐  win the day!

**Pull request merged**